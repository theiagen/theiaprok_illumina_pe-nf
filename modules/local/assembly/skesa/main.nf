process SKESA {
    tag "$meta.id"
    label "process_medium"
    
    container "us-docker.pkg.dev/general-theiagen/staphb/skesa:2.4.0"

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path('*_contigs.fa'), emit: assembly_fasta
    path "versions.yml"                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def min_contig_len = params.min_contig_length ?: 1
    def maxmem = task.memory.toGiga()
    def reads_input = meta.single_end ? "--fastq ${reads[0]}" : "--fastq ${reads[0]},${reads[1]} --use_paired_ends"
    
    """
    skesa \\
        --gz \\
        ${reads_input} \\
        --contigs_out ${prefix}_skesa_contigs.fa \\
        --min_contig ${min_contig_len} \\
        --memory ${maxmem} \\
        --cores ${task.cpus} \\
        --vector_percent 1 \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        skesa: \$(skesa --version 2>&1 | grep "SKESA" | sed -E 's/^.*SKESA ([0-9.]+).*/\\1/')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_skesa_contigs.fa

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        skesa: \$(skesa --version 2>&1 | grep "SKESA" | sed -E 's/^.*SKESA ([0-9.]+).*/\\1/')
    END_VERSIONS
    """
}