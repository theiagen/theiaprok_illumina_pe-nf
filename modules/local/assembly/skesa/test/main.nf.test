nextflow_process {
    name "Test Process SKESA"
    script "../main.nf"
    process "SKESA"
    
    test("skesa - test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.assembly_fasta,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("skesa - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}