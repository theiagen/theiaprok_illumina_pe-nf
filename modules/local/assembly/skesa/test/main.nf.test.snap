{
    "skesa - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_skesa_contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5ff4a21d350369c0212a4b22678a252"
                ],
                "assembly_fasta": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_skesa_contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5ff4a21d350369c0212a4b22678a252"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T15:07:55.26144307"
    },
    "skesa - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_skesa_contigs.fa:md5,409b15cbfbc8fa170152004fb5c59a31"
                ]
            ],
            [
                "versions.yml:md5,a5ff4a21d350369c0212a4b22678a252"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T15:07:42.212845223"
    }
}